tests/baselines/reference/unionTypeCallSignatures5.errors.txt

```diff
@@ -0,0 +1,19 @@
+tests/cases/conformance/types/union/unionTypeCallSignatures5.ts(12,1): error TS2684: The 'this' context of type 'void' is not assignable to method's 'this' of type 'never'.
+
+
+==== tests/cases/conformance/types/union/unionTypeCallSignatures5.ts (1 errors) ====
+    // #31485
+    interface A {
+      (this: void, b?: number): void;
+    }
+    interface B {
+      (this: number, b?: number): void;
+    }
+    interface C {
+      (i: number): void;
+    }
+    declare const fn: A | B | C;
+    fn(0);
+    ~~~~~
+!!! error TS2684: The 'this' context of type 'void' is not assignable to method's 'this' of type 'never'.

```

Comment:
I’m not sure what to do with `this: void`. It doesn’t have any special handling right now, so it intersects with `number` to produce `never`. But it feels fairly synonymous with completely omitting a `this` type, in which case I would drop it out of the intersection. I’m a bit fuzzy on how to interpret the `void` type outside of return types.

Reply:
`void` is _usually_ best treated as synonymous with `undefined` in all non-return positions - at least that's what we've done thus far (despite the fact that any return type is assignable to a void returning function, thus you cannot guarantee that a void returning function returns `undefined`). It's a bit inconsistent.

Reply:
Ok; by that logic I think the intersection stands.

Reply:
In this parameters, `void` is *definitely* a synonym of `undefined`. I wrote the feature before `undefined` existed. `this: void` is (was?) the idiomatic way to specify that there will be no `this` provided. 

`never` seems right because it arises from an union of signatures that require both (1) no `this` provided (2) some `this` provided.